feat: Add SOCKS5 Proxy support (for Tor)

Milor123 opened the pull request:
I'd like to propose a new feature that I think could be very useful. This pull request adds support for using a **SOCKS5 proxy**.

The main goal was to enable the plugin to be used through **Tor** (which typically uses a SOCKS5 proxy on port 9050), adding an extra layer of privacy. The existing HTTP proxy works great, and this just opens up more possibilities. 🔒

<img width="828" height="666" alt="imagen" src="https://github.com/user-attachments/assets/b4383242-0396-47aa-b858-51a724250822" />

**A Note on Development:**

To implement this, I had help from Jules, Google's code agent. While the AI was a huge help in drafting the code, I made sure to manually test every single change. We went through several iterations (and a few funny bugs! xD 😄) to ensure that this new functionality doesn't break the existing HTTP proxy or the normal translation process.

**What was done in a nutshell:**

*   ✅ Added a new section in the settings for the SOCKS5 proxy, with its own connection test button.
*   ⚙️ Integrated the `PySocks` library to handle SOCKS connections safely and locally, only when they are needed.
*   🧪 Verified that the connection tests and the main translation feature work correctly with HTTP, SOCKS5, and with no proxy enabled.

As I'm not developer, I would be extremely grateful if you could **please review the code carefully**. We just want to be sure the implementation is solid and won't cause any unexpected side effects in other parts of the plugin.

Thank you so much for your time and for maintaining this amazing add-on! 🙏

bookfere commented:
Thank you for your contribution and for adding such a useful feature to the plugin! I have carefully reviewed your code and found that the network proxy settings from the UI can be merged together. I made some refactoring and ensured that all test cases pass. Please review and test the updates, and feel free to leave any feedback.

Milor123 commented:
> Thank you for your contribution and for adding such a useful feature to the plugin! I have carefully reviewed your code and found that the network proxy settings from the UI can be merged together. I made some refactoring and ensured that all test cases pass. Please review and test the updates, and feel free to leave any feedback.

Thanks man!

Here is working for both HTTP and SOCK5
<img width="830" height="698" alt="imagen" src="https://github.com/user-attachments/assets/ce4c2dc3-8e73-4651-9e2a-4ea27b6e2c21" />


but when try do the translate using SOCK5 i get this and the program stucks  (I have similar issue while try solve it using Jules)

<img width="606" height="597" alt="imagen" src="https://github.com/user-attachments/assets/2d019b44-12d4-48da-85f2-cfc5c767a769" />


```
calibre, version 8.8.0
ERROR: Excepción no considerada: <b>GeneralProxyError</b>:Socket error: 0x01: General SOCKS server failure

calibre 8.8  embedded-python: True
Windows-10-10.0.26100-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.26100')
Python 3.11.12
Windows: ('10', '10.0.26100', 'SP0', 'Multiprocessor Free')
Interface language: es
EXE path: C:\Program Files\Calibre2\calibre.exe
Successfully initialized third party plugins: Ebook Translator (2, 4, 1)
Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.lib.handler", line 16, in __init__
  File "asyncio\events.py", line 681, in get_event_loop
RuntimeError: There is no current event loop in thread 'Dummy-7'.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.lib.socks", line 808, in connect
  File "calibre_plugins.ebook_translator.lib.socks", line 443, in _negotiate_SOCKS5
  File "calibre_plugins.ebook_translator.lib.socks", line 533, in _SOCKS5_request
calibre_plugins.ebook_translator.lib.socks.SOCKS5Error: 0x01: General SOCKS server failure

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.advanced", line 213, in translate_paragraphs
  File "calibre_plugins.ebook_translator.lib.translation", line 227, in handle
  File "calibre_plugins.ebook_translator.lib.handler", line 18, in __init__
  File "asyncio\events.py", line 810, in new_event_loop
  File "asyncio\events.py", line 699, in new_event_loop
  File "asyncio\selector_events.py", line 56, in __init__
  File "asyncio\selector_events.py", line 107, in _make_self_pipe
  File "socket.py", line 621, in _fallback_socketpair
  File "calibre_plugins.ebook_translator.lib.socks", line 47, in wrapper
  File "calibre_plugins.ebook_translator.lib.socks", line 813, in connect
calibre_plugins.ebook_translator.lib.socks.GeneralProxyError: Socket error: 0x01: General SOCKS server failure

```

and TOR saiys:

```
Aug 21 17:10:43.000 [warn] Rejecting SOCKS request for anonymous connection to private address [scrubbed].
```

The HTTP mode works ✅ 
The SOCK5 Fail ❌ 

I am not developer, I dont know how fix.

Milor123 commented:
> There was a conflict between the HTTP proxy and the SOCKS proxy. I have fixed the issue. Please give it a try and let me know if you encounter any problems.
> 
> BTW, I used the "Tor Browser," which has a built-in SOCKS5 proxy, to test this feature.

Good morning bro!

I've tried using TOR and TOR browser, (also tried disable my firewall and network things of eset nod32)

## Tor-Browser test (firewall off / eset off)
<img width="494" height="376" alt="imagen" src="https://github.com/user-attachments/assets/06579414-65f4-4825-a702-455034830914" />

```
calibre, version 8.9.0
ERROR: Unhandled exception: <b>ProxyConnectionError</b>:Error connecting to SOCKS5 proxy 127.0.0.1:9050: [WinError 10061] No se puede establecer una conexión ya que el equipo de destino denegó expresamente dicha conexión

calibre 8.9  embedded-python: True
Windows-10-10.0.26100-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.26100')
Python 3.11.12
Windows: ('10', '10.0.26100', 'SP0', 'Multiprocessor Free')
Interface language: None
EXE path: C:\Program Files\Calibre2\calibre.exe
Successfully initialized third party plugins: Ebook Translator (2, 4, 1)
Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.lib.handler", line 16, in __init__
  File "asyncio\events.py", line 681, in get_event_loop
RuntimeError: There is no current event loop in thread 'Dummy-7'.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.lib.socks", line 787, in connect
ConnectionRefusedError: [WinError 10061] No se puede establecer una conexión ya que el equipo de destino denegó expresamente dicha conexión

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.advanced", line 213, in translate_paragraphs
  File "calibre_plugins.ebook_translator.lib.translation", line 227, in handle
  File "calibre_plugins.ebook_translator.lib.handler", line 18, in __init__
  File "asyncio\events.py", line 810, in new_event_loop
  File "asyncio\events.py", line 699, in new_event_loop
  File "asyncio\selector_events.py", line 56, in __init__
  File "asyncio\selector_events.py", line 107, in _make_self_pipe
  File "socket.py", line 621, in _fallback_socketpair
  File "calibre_plugins.ebook_translator.lib.socks", line 47, in wrapper
  File "calibre_plugins.ebook_translator.lib.socks", line 800, in connect
calibre_plugins.ebook_translator.lib.socks.ProxyConnectionError: Error connecting to SOCKS5 proxy 127.0.0.1:9050: [WinError 10061] No se puede establecer una conexión ya que el equipo de destino denegó expresamente dicha conexión
```

## TOR cli test (firewall off / eset off)

<img width="494" height="352" alt="imagen" src="https://github.com/user-attachments/assets/f9af06aa-728f-4495-a9df-6aecdb41b575" />
<img width="1105" height="392" alt="imagen" src="https://github.com/user-attachments/assets/a0e64f9c-af5e-4454-b279-4566fe6dcee2" />


## My think

Maybe you might think it’s just some misconfiguration in my network rules or something similar, but that wouldn’t make sense—look:

### Translate in Socks5

<img width="828" height="666" alt="image" src="https://github.com/user-attachments/assets/b9c7f92f-c890-4790-8c24-0dd543b5a507" />

In the image above it works perfectly. (In fact, if I kill my Tor CLI process and then try to translate, it fails.)

That’s why I believe the problem must lie in the book’s translation module. I should also mention that when I tested the socks5h version of the code that Jules built, I didn’t have to disable my antivirus, firewall, or any network settings—it just worked without any issues.

bookfere commented:
For "Tor Browser," you should use port `9150`.

Regarding your question, I did not implement anything new except refactoring your code for maintainability. I do not quite understand your concern about SOCKS5 vs. SOCKS5h. Aren’t they both handled by the library PySocks you added?

I have tested the feature using both "Tor Browser" and "Tor CLI," and they worked well. Since you suspect the problem is related to the translation module, I suggest running Calibre in debug mode:

```
cd /PATH/TO/THE/LATEST/CODE
calibre-customize -b .; calibre-debug -sg
```

Then open the Ebook Translator and translate a paragraph. If you can see that the socket in use is the same as the patched socket, it proves that everything is working normally.

<img width="423" height="150" alt="sockets-proxy" src="https://github.com/user-attachments/assets/7f3e64c7-e85e-493c-8953-dbf4009f2115" />

Milor123 commented:
<img width="1233" height="656" alt="imagen" src="https://github.com/user-attachments/assets/7d006d0f-bee4-444d-9af3-5ecf68de6a07" />

Same issue bro, in debug mode, TOR CLI

```
Windows-10-10.0.26100-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.26100')
Python 3.11.12
Windows: ('10', '10.0.26100', 'SP0', 'Multiprocessor Free')
Interface language: None
EXE path: C:\Program Files\Calibre2\calibre-debug.exe
Successfully initialized third party plugins: Ebook Translator (2, 4, 1)
Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.lib.handler", line 16, in __init__
  File "asyncio\events.py", line 681, in get_event_loop
RuntimeError: There is no current event loop in thread 'Dummy-7'.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.lib.socks", line 809, in connect
  File "calibre_plugins.ebook_translator.lib.socks", line 443, in _negotiate_SOCKS5
  File "calibre_plugins.ebook_translator.lib.socks", line 533, in _SOCKS5_request
calibre_plugins.ebook_translator.lib.socks.SOCKS5Error: 0x01: General SOCKS server failure

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.advanced", line 213, in translate_paragraphs
  File "calibre_plugins.ebook_translator.lib.translation", line 227, in handle
  File "calibre_plugins.ebook_translator.lib.handler", line 18, in __init__
  File "asyncio\events.py", line 810, in new_event_loop
  File "asyncio\events.py", line 699, in new_event_loop
  File "asyncio\selector_events.py", line 56, in __init__
  File "asyncio\selector_events.py", line 107, in _make_self_pipe
  File "socket.py", line 621, in _fallback_socketpair
  File "calibre_plugins.ebook_translator.lib.socks", line 47, in wrapper
  File "calibre_plugins.ebook_translator.lib.socks", line 814, in connect
calibre_plugins.ebook_translator.lib.socks.GeneralProxyError: Socket error: 0x01: General SOCKS server failure
```

<img width="1149" height="394" alt="imagen" src="https://github.com/user-attachments/assets/79d13ce8-adfe-4b42-b91d-7f0a49524d16" />

bookfere commented:
I tested the revision on macOS without issues, but on Windows it failed. You are right—the problem was in the Translation module: the patched socket was interfering with the asyncio routines. I’ve just fixed it; please give it a try.

bookfere commented:
Great! Let's merge it. Feel free to share any more thoughts.

Milor123 commented:
@bookfere I'm afraid I’ve found a bug: the proxy won’t turn off—the proxy option stays checked, so I’m forced to use HTTP or SOCKS and can’t disable them. Strange, isn’t it? Doesn’t the same thing happen to you?

Milor123 commented:
Well, apparently I can disable it; however, I can no longer do batch translation. When I press batch translation, this happens

```

calibre, version 8.10.0
ERROR: Unhandled exception: <b>KeyError</b>:''

calibre 8.10  embedded-python: True
Windows-10-10.0.26100-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.26100')
Python 3.11.12
Windows: ('10', '10.0.26100', 'SP0', 'Multiprocessor Free')
Interface language: None
EXE path: C:\Program Files\Calibre2\calibre.exe
Successfully initialized third party plugins: Ebook Translator (2, 4, 1)
Traceback (most recent call last):
  File "calibre_plugins.ebook_translator.ui", line 170, in select_preferred_mode
  File "calibre_plugins.ebook_translator.ui", line 100, in show_batch_translation
  File "calibre_plugins.ebook_translator.batch", line 38, in __init__
  File "calibre_plugins.ebook_translator.batch", line 42, in main_layout
  File "calibre_plugins.ebook_translator.batch", line 137, in layout_translate
  File "calibre_plugins.ebook_translator.batch", line 135, in change_target_lang
  File "calibre_plugins.ebook_translator.engines.base", line 99, in get_iso639_target_code
  File "calibre_plugins.ebook_translator.engines.base", line 95, in get_target_code
KeyError: ''

```

At first I thought it was because I had added some custom GPTs a few days ago. So I switched to the Microsoft translator, saved the changes, but for some reason the function to translate the default batch I have is broken.

Is this the problem from here?

Milor123 commented:
> Thanks for catching the bug! I’ve made a fix—please give it another try :)
It seems to be working and the batch-processing error has been fixed. Thank you very much.
This week, if I can, I’ll run a more thorough test to see if I can break anything.
